RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,823 @@
+
+// Array type signature
+let f (arr: (float | null) []) = ()
+
+// Nullable types with nullable generic types
+type C<'T | null>() =
+    member __.M(param1: List<List<'T | null> | null> | null) = ()
+```
+
+The syntax accomplishes two goals:
+
+1. Makes the distinction that a nullable reference type can be `null` clear in syntax.
+2. Becomes a bit unweildy when nullability is used a lot, particularly with nested types.
+
+Because it is a design goal to flow **non-null** reference types, and **avoid** flowing nullable reference types unless absolutely necessary, it is considered a feature that the syntax can get a bit unweildy if used everywhere. That is, we want to discourage the use of `null` unless it is absolutely necessary.
+
+#### Nullable reference type declarations in F\#
+
+There will be no way to declare an F# type as follows:
+
+```fsharp
+// This will not be possible!
+type (SomeClass | null)() = class end
+```
+
+And in existing F# today, the only way to declare an F# reference type that could be `null` in F# code is to use `[<AllowNullLiteral>]`. This would not change with this proposal; that is, the only way to declare a reference type in F# that could have `null` as a value in F# code is `[<AllowNullLiteral>]`.
+
+### Checking of nullable references
+
+There are a few common ways that F# programmers check for null today. Unless explicitly mentioned as unsupported, support for flow analysis is under consideration for certain null-checking patterns.
+
+#### Likely supported: Pattern matching with alias
+
+```fsharp
+let len (str: string | null) =
+    match str with
+    | null -> -1
+    | s -> s.Length // OK - we know 's' is string
+```
+
+This is by far the most common pattern in F# programming. In the previous code sample, `str` is of type `string | null`, but `s` is now of type `string`. This is because we know that based on the `null` has been accounted for by the `null` pattern.
+
+#### Likely supported: isNull and others using null-handling decoration
+
+```fsharp
+let len (str: string | null) =
+    if isNull str then
+        -1
+    else
+        str.Length // OK
+```
+
+The `isNull` function is considered to be a good way to check for `null`. We can annotate it in FSharp.Core with the attribute used for functions handling null (we'll call it `[<HandlesNull>]`).
+
+Similarly, CoreFX will annotate methods like `String.IsNullOrEmpty` as handling `null`. This means that the most common of null-checking methods can be respected by flow analysis and reduce the amount of nullability assertions.
+
+##### Possibly supported: Pattern matching with wildcard
+
+```fsharp
+let len (str: string | null) =
+    match str with
+    | null -> -1
+    | _ -> str.Length // OK - 'str' is string
+```
+
+A slight riff that is less common is treating `str` differently when in a wildcard pattern's scope. To reach that code path, `str` must indeed by non-null, so this could potentially be supported.
```

Comment:
My full opinion is that this should not really be supported; that is, you need to give it a name. But I do think that this may be unexpected by users, even if it is consistent with how things are done today. If we don't do this, we may want to consider a special warning suggesting that you give it an alias if you intend on using `str` as if it were a non-nullable string.